source/extensions/filters/listener/tls_inspector/ja4_fingerprint.cc

```diff
@@ -0,0 +1,324 @@
+#include "source/extensions/filters/listener/tls_inspector/ja4_fingerprint.h"
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "openssl/evp.h"
+#include "openssl/sha.h"
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace TlsInspector {
+
+namespace {
+// Google GREASE values as defined in RFC 8701
+// These values are used to prevent ossification in TLS extensions and should be
+// ignored when fingerprinting clients to prevent artificial differentiation
+// See: https://datatracker.ietf.org/doc/html/rfc8701
+const std::array<uint16_t, 16> GREASE_VALUES = {
+    0x0a0a, 0x1a1a, 0x2a2a, 0x3a3a, 0x4a4a, 0x5a5a, 0x6a6a, 0x7a7a,
+    0x8a8a, 0x9a9a, 0xaaaa, 0xbaba, 0xcaca, 0xdada, 0xeaea, 0xfafa,
+};
+} // namespace
+
+bool JA4Fingerprinter::isNotGrease(uint16_t id) {
+  return std::find(GREASE_VALUES.begin(), GREASE_VALUES.end(), id) == GREASE_VALUES.end();
+}
+
+absl::string_view JA4Fingerprinter::mapVersionToString(uint16_t version) {
+  switch (version) {
+  case TLS1_3_VERSION:
+    return "13";
+  case TLS1_2_VERSION:
+    return "12";
+  case TLS1_1_VERSION:
+    return "11";
+  case TLS1_VERSION:
+    return "10";
+  default:
+    return "00";
+  }
+}
+
+absl::string_view JA4Fingerprinter::getJA4TlsVersion(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  const uint8_t* data;
+  size_t size;
+  if (SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_supported_versions, &data,
+                                           &size)) {
+    CBS cbs;
+    CBS_init(&cbs, data, size);
+    CBS versions;
+
+    if (!CBS_get_u8_length_prefixed(&cbs, &versions)) {
+      return "00";
+    }
+
+    uint16_t highest_version = 0;
+    uint16_t version;
+    while (CBS_get_u16(&versions, &version)) {
+      if (isNotGrease(version) && version > highest_version) {
+        highest_version = version;
+      }
+    }
+
+    if (highest_version != 0) {
+      return mapVersionToString(highest_version);
+    }
+  }
+
+  // Fallback to the protocol version if the supported_versions extension is not present
+  return mapVersionToString(ssl_client_hello->version);
+}
+
+bool JA4Fingerprinter::hasSNI(const SSL_CLIENT_HELLO* ssl_client_hello) {
+  const uint8_t* data;
+  size_t size;
+  return SSL_early_callback_ctx_extension_get(ssl_client_hello, TLSEXT_TYPE_server_name, &data,
+                                              &size);
+}
+
+int JA4Fingerprinter::countCiphers(const SSL_CLIENT_HELLO* ssl_client_hello) {
```

Comment:
uint32_t instead of int? I don't think this should be signed, and sized types are preferred.